homeassistant/components/mqtt/client.py

```diff
@@ -456,25 +468,140 @@ def cleanup(self) -> None:
         while self._cleanup_on_unload:
             self._cleanup_on_unload.pop()()
 
+    @contextlib.asynccontextmanager
+    async def _async_connect_in_executor(self) -> AsyncGenerator[None, None]:
```

Comment:
I made this a context manager so its much harder to get wrong.  Any time we call connect/reconnect in the executor we need to wrap it in this to make sure the callbacks are calling the `call_soon_threadsafe` versions when the executor job is running